### PR TITLE
Irq combine and add some function descriptions

### PR DESCRIPTION
**Pull request by hannahhoang2704**
This branch is developed based on roni's branch but I'll put pull request and compare with dev-master. It combine irq callback event from both GPIO opto fork and piezo sensor

### CODE REVIEW

**Comment by RoniNuutinen** [comment]
I'm finding it somewhat convoluted how the flags are operated now...
- In rotate_to_event the opto_flags are set with set_opto_flag, while being externalized for direct handling aswell for the irq_handler...
- And the piezo_flag is handled directly as the variable, also 'externally', though still operated in dispense(..) directly as the variable instead of with a function like opto is...

I'm guessing you preferred leaving the flags in opto.c and piezo.c respectively as they pertain to those pins - kinda like macros -- I get that.
.. But I find the externalization and inconsistency obscuring...

Alternatively:
- the flags and "set_flag" -functions could be declared in operation.c
removing the need for externalization, thus clarifying their handling in my opinion.
- I would prefer that opto and peizo flags were operated in the same way throughout the code, as in:
either with the set_flag -functions,
or directly as the variables,
for consistency's sake,
as opposed to how they are currently operated via functions **and** directly atm..


I can write line-directed comments if this comment is unclear. They just are throughout the code in a few different places making it harder to combine them as a single coherent suggestion...

**Comment by RoniNuutinen** [comment]
Also,
moving the flag declarations to operation.c would eliminate the need for set_flag -functions altogether,
as flag manipulation is needed only in operation.c,
skimming the code down in general.

**Comment by RoniNuutinen** [comment]
Do you disagree ; am I being too critical?

**Comment by hannahhoang2704** [comment]
Not sure if I understand your comment well, so all the functions and flags from piezo and opto are now in operation.c. no extern flags. Still I think set_piezo_flag function is unnecessary since it can be set directly without a need of creating a separate function in operation.c to set the flag to false while in set_opto_flag function you have 2 separate flags and depending on what you would like to detect (rise or fall) you set the different flag. anyway all functions are now in operation.c
if you can come up with any idea to simplify that opto flag that would be also great

**Comment by RoniNuutinen** [comment]
Oh, right... I forgot the rotate_to_opto setup...

I do have an idea for skimming it cleaner, which is difficult to explain.
I can try that after having merged some version of this into dev-master.

**Comment by RoniNuutinen** [comment]
And so how do we proceed with our two PR's? I feel like we've been stuck for some time now and as if this is escalating...
Shall I press this green button here now,
after which I or you resolve the conflict with 'roni PR',
or...?

**Comment by hannahhoang2704** [comment]
I'm now waiting for Phuong's review, if he won't give any more comment by sunday I'll merge both of these branches to dev-master. I can handle the merge on Sunday evening

**Comment by RoniNuutinen** [comment]
Ok.
Would you like me to try and write my idea of the irq-flag handling I mentioned above to irq-combine?
I do have a fairly clear and confident idea.
... or should I just wait?

**Comment by hannahhoang2704** [comment]
yes you can pull this branch to your local, do the adjust on and push to this same branch remotely. so we don't need to wait for merging.

**Comment by RoniNuutinen** [comment]
I left the functions still in, just commented out, in case there's some need to revert back.

File: operation/operation.c [hunk 1]
```diff
@@ -5,9 +5,9 @@
 
 #include "operation.h"
 
-static volatile bool opto_fall = false;
-static volatile bool opto_rise = false;
-static volatile bool piezo_falling_edge = false;
+static volatile bool interrupt_flags[] = {
+        false, false, false
+};
```
Comment (hannahhoang2704) [inline operation/operation.c]: this is cleaner than before. I think struct would give clearer structure than array. then you don't need enum either. I would prefer using struct.

Comment (RoniNuutinen) [reply operation/operation.c]: Hmm... I'm unsure how they would be called/handled without enum ;
removing the enum would require us to call the flags with 0, 1 and 2 -- which works, but isn't very clear ; hence the enum declaration.

I'm interested if you can do it without enums, but am unsure how struct could replace them.

Comment (hannahhoang2704) [reply operation/operation.c]: yeah, you're right. struct in this case maybe wont make it clear enough. Let's go with this!

File: eeprom/eeprom.c [hunk 2]
```diff
@@ -125,5 +119,6 @@ void erase_logs(uint8_t *log_index) {
         write_to_eeprom(write_address, buf, 1);
     }
     *log_index = 0;
+    write_to_eeprom(LOG_INDEX_ADDR, log_index, 1);
```
Comment (hannahhoang2704) [inline eeprom/eeprom.c:122]: store a new log index to EEPROM

File: operation/operation.c [hunk 3]
```diff
@@ -123,7 +89,6 @@ void logf_msg(enum logs logEnum, oper_st * state, int n_args, ...) {
 
     printf("%s\n", msg);
     write_log_entry(msg, &state->eeprom_log_idx);
-    write_to_eeprom(LOG_INDEX_ADDR, &state->eeprom_log_idx, 1);
```
Comment (hannahhoang2704) [inline operation/operation.c]: line 126 is removed since write_log_entry() function do store the log index to EEPROM, no need to call it again here

**Review by hannahhoang2704 — COMMENTED** [review]
A small change related to store log index in EEPROM

File: operation/operation.c [hunk 4]
```diff
@@ -127,6 +127,7 @@ oper_st init_operation() {
     oper_st state;
 
     state.eeprom_log_idx = get_stored_value(LOG_INDEX_ADDR);
+    if(state.eeprom_log_idx >= MAX_ENTRIES) state.eeprom_log_idx = 0;
```
Comment (hannahhoang2704) [inline operation/operation.c:127]: add this to make sure if eeprom hasn't stored any log index then eeprom_log_idx  is within entries size

File: operation/operation.c [hunk 5]
```diff
@@ -165,6 +166,7 @@ void blink_until_sw_pressed(SW * sw_proceed, LED * led) {
         loop %= 10;
         sleep_ms(50);
     }
+    set_led_brightness(led, PWM_OFF);
```
Comment (hannahhoang2704) [inline operation/operation.c]: if button is pressed when the led is toggle on, the led will be on during calibration, line 169 is to be sure that the light is off after button is pressed

**Review by hannahhoang2704 — COMMENTED** [review]
Just a minor change

File: Lora/Lora.c [hunk 6]
```diff
@@ -33,14 +33,13 @@ void start_lora() {
     const uint8_t init_command[] = "AT\r\n";
     char *receivedString;
     for (size_t i = 0; i < 5; i++) {
-        uart_write_blocking(UART_ID, init_command, strlen(init_command));
+        uart_write_blocking(UART_ID, init_command, strlen((const char*)init_command));
```
Comment (hannahhoang2704) [inline Lora/Lora.c:36]: typecasting to avoid warning (init_command is uint8_t type and strlen() function only works with char type

File: operation/operation.c [hunk 7]
```diff
@@ -110,7 +110,7 @@ void logf_msg(enum logs logEnum, oper_st * state, int n_args, ...) {
             write_to_eeprom(COMP_INDEX_ADDR,
                             &state->current_comp_idx, 1);
             if (state->lora_connected)
-                send_msg_to_Lora(msg, false);
+                send_msg_to_Lora(msg, true);
```
Comment (hannahhoang2704) [inline operation/operation.c]: One thing I'm not sure about is whether we need to get response from Lora after sending message? Is it in our requirement somehow or this is just our own decision that we want to sometimes check if we get response from lora when sending command?

**Review by hannahhoang2704 — COMMENTED** [review]
small comment and question related to Lora

**Comment by RoniNuutinen** [comment]
Many of these few last changes and mentions have been addressed in the 'roni -> dev-master' PR,
as that one pertains to the EEPROM and LoRa implementation.

Whether LoRa `msg` responses need or need not be confirmed needs to be tested.
I applied the `wait_for_rsp` option originally for testing purposes,
to confirm that the `msg` is not just blindly sent but also acknowledged as received by the module.
I however set it to `false` for logically time-sensitive sections such as the rotation + pill detection, as waiting for response adds a several-second delay.
During implementation, as I had several new things implemented, the program sometimes got stuck in a very obscure way which I can't explain, and one thing I tried was setting `wait_for_rsp` to `true` for those `msg`'s: it stopped getting stuck --- however, I might have made other changes as well, I don't remember; as I said I had several balls in the air.

If I recall correctly, it got stuck at `ROTATION_COMPLETED` -> `PILL_DETECTED` (or not detected), where the two `msg`'s were sent back to back, without delay (when not waiting for response).

A parallel logic would be the lora connection operation:
It fails if the response isn't waited for, because each command needs to succeed for the following commands to possibly work.
**Possibly**, if a `msg` is sent to the module and the module hasn't finished processing a possible previous `msg`,
maybe, it can get stuck somehow.

TL;DR: Waiting for response for LoRa `msg`'s might be redundant -- although required for connection establishment. I don't think the project assignment requires us to wait for response.
The delay is annoying, which is why waiting for `msg` responses would preferably be set off -- but it might be required for the program to proceed reliably.

**Comment by hannahhoang2704** [comment]
maybe we should check more Lora with Phuong, he explain somehow that his new code is shorten the waiting time of response
